internal/backends/sqlite/collection.go

```diff
@@ -91,6 +90,156 @@ func (c *collection) Query(ctx context.Context, params *backends.QueryParams) (*
+
+	for {
+		k, v, err := iter.Next()
+		if errors.Is(err, iterator.ErrIteratorDone) {
+			break
+		}
+
+		if err != nil {
+			return "", nil, lazyerrors.Error(err)
+		}
+
+		// queryPath stores the path that is used in SQLite to access specific key
+		// if the key is _id we use our predifined path, as the handling of _id may
+		// change in the future
+		queryPath := metadata.IDColumn
+
+		// keyArgs store the optional parameters used to query the key
+		var keyArgs []any
+
+		if k != "_id" {
+			// To use parameters inside of SQLite json path the parameter token ("?")
+			// needs to be concatenated to path with || operator
+			queryPath = fmt.Sprintf(`%s->('$."' || ? || '"' )`, metadata.DefaultColumn)
+			keyArgs = append(keyArgs, k)
+		}
+
+		// don't pushdown $comment
+		if strings.HasPrefix(k, "$") {
+			continue
+		}
+
+		path, err := types.NewPathFromString(k)
+
+		var pe *types.PathError
+
+		switch {
+		case err == nil:
+			// TODO https://github.com/FerretDB/FerretDB/issues/2069
+			if path.Len() > 1 {
+				continue
+			}
+		case errors.As(err, &pe):
+			// ignore empty key error, otherwise return error
+			if pe.Code() != types.ErrPathElementEmpty {
+				return "", nil, lazyerrors.Error(err)
+			}
+		default:
+			panic("Invalid error type: PathError expected")
+		}
+
+		switch v := v.(type) {
+		case *types.Document, *types.Array, types.Binary, types.NullType, types.Regex, types.Timestamp:
+			// type not supported for pushdown
+			continue
+
+		case float64:
+			var comparison string
+
+			switch {
+			case v > types.MaxSafeDouble:
+				comparison = ` > ?`
+				v = types.MaxSafeDouble
+
+			case v < -types.MaxSafeDouble:
+				comparison = ` < ?`
+				v = -types.MaxSafeDouble
+			default:
+				// don't change the default eq query
+			}
+
+			subquery := fmt.Sprintf(`EXISTS (SELECT value FROM json_each(%v) WHERE value %s)`, queryPath, comparison)
+			filters = append(filters, subquery)
+
+			// TODO https://github.com/FerretDB/FerretDB/issues/3386
+			args = append(args, keyArgs...)
+			args = append(args, parseValue(v))
+
+		case types.ObjectID, time.Time, string, bool, int32:
+			subquery := fmt.Sprintf(`EXISTS (SELECT value FROM json_each(%v) WHERE value = ?)`, queryPath)
+			filters = append(filters, subquery)
+
+			// TODO https://github.com/FerretDB/FerretDB/issues/3386
+			args = append(args, keyArgs...)
+			args = append(args, parseValue(v))
+
+		case int64:
+			var comparison string
+			maxSafeDouble := int64(types.MaxSafeDouble)
+
+			// If value cannot be safe double, fetch all numbers out of the safe range
+			switch {
+			case v > maxSafeDouble:
+				comparison = ` > ?`
+				v = maxSafeDouble
+
+			case v < -maxSafeDouble:
+				comparison = `< ?`
+				v = -maxSafeDouble
+			default:
+				// don't change the default eq query
+			}
+
+			// json_each returns top level json values, and the contents of arrays if any
+			subquery := fmt.Sprintf(`EXISTS (SELECT value FROM json_each(%v) WHERE value %s)`, queryPath, comparison)
```

Comment:
By the way, why do we use `%v` in `json_each(%v)`?

Reply:
fixed